plugin/storage/badger/factory.go

```diff
@@ -0,0 +1,179 @@
+
+package badger
+
+import (
+	"expvar"
+	"flag"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"github.com/dgraph-io/badger"
+	"github.com/spf13/viper"
+	"github.com/uber/jaeger-lib/metrics"
+	"go.uber.org/zap"
+
+	badgerStore "github.com/jaegertracing/jaeger/plugin/storage/badger/spanstore"
+	"github.com/jaegertracing/jaeger/storage/dependencystore"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+var (
+	// ValueLogSpaceAvailable returns the amount of space left on the value log mount point in bytes
+	ValueLogSpaceAvailable *expvar.Int
+	// KeyLogSpaceAvailable returns the amount of space left on the key log mount point in bytes
+	KeyLogSpaceAvailable *expvar.Int
+	// LastMaintenanceRun stores the timestamp of the previous maintenanceRun
+	LastMaintenanceRun *expvar.Int
+	// LastValueLogCleaned stores the timestamp of the previous ValueLogGC run
+	LastValueLogCleaned *expvar.Int
+)
+
+// Factory implements storage.Factory for Badger backend.
+type Factory struct {
+	Options *Options
+	store   *badger.DB
+	cache   *badgerStore.CacheStore
+	logger  *zap.Logger
+
+	tmpDir              string
+	maintenanceInterval time.Duration
+	maintenanceTicker   *time.Ticker
+}
+
+const (
+	defaultTickerInterval time.Duration = 5 * time.Minute
+)
+
+// NewFactory creates a new Factory.
+func NewFactory() *Factory {
+	if ValueLogSpaceAvailable == nil {
+		ValueLogSpaceAvailable = expvar.NewInt("badger_value_log_bytes_available")
+	}
+	if KeyLogSpaceAvailable == nil {
+		KeyLogSpaceAvailable = expvar.NewInt("badger_key_log_bytes_available")
+	}
+	if LastMaintenanceRun == nil {
+		LastMaintenanceRun = expvar.NewInt("badger_storage_maintenance_last_run")
+	}
+	if LastValueLogCleaned == nil {
+		LastValueLogCleaned = expvar.NewInt("badger_storage_valueloggc_last_run")
+	}
+	return &Factory{
+		Options:             NewOptions("badger"),
+		maintenanceInterval: defaultTickerInterval,
+	}
+}
+
+// AddFlags implements plugin.Configurable
+func (f *Factory) AddFlags(flagSet *flag.FlagSet) {
+	f.Options.AddFlags(flagSet)
+}
+
+// InitFromViper implements plugin.Configurable
+func (f *Factory) InitFromViper(v *viper.Viper) {
+	f.Options.InitFromViper(v)
+}
+
+// Initialize implements storage.Factory
+func (f *Factory) Initialize(metricsFactory metrics.Factory, logger *zap.Logger) error {
+	f.logger = logger
+
+	opts := badger.DefaultOptions
+
+	if f.Options.primary.Ephemeral {
+		opts.SyncWrites = false
+		// Error from TempDir is ignored to satisfy Codecov
+		dir, _ := ioutil.TempDir("", "badger")
+		f.tmpDir = dir
+		opts.Dir = f.tmpDir
+		opts.ValueDir = f.tmpDir
+
+		f.Options.primary.KeyDirectory = f.tmpDir
+		f.Options.primary.ValueDirectory = f.tmpDir
+	} else {
+		opts.SyncWrites = f.Options.primary.SyncWrites
+		opts.Dir = f.Options.primary.KeyDirectory
+		opts.ValueDir = f.Options.primary.ValueDirectory
+	}
+
+	store, err := badger.Open(opts)
+	if err != nil {
+		return err
+	}
+	f.store = store
+
+	// Error ignored to satisfy Codecov
```

Comment:
The error comes from the internal logic, is there no way to simulate it? If we're always ignoring it, is there a reason to even return it?

Reply:
The error could come from corrupted storage / other issues in database files. I think these should be returned, but making badger fail to load LSM tree for tests is quite beyond the scope of Jaeger I would say. But codecov does not allow returning it (I'm sure user would be happy to know that something went wrong).